composer/utils/checkpoint.py

```diff
@@ -443,23 +453,75 @@ def __init__(self, source_path: str, destination_path: str, object_store):
-                                                      filename=file_destination)
+            device_mesh = state.fsdp_device_mesh
+            first_replica = device_mesh is None or device_mesh.get_local_rank(mesh_dim=0) == 0
+
+            # 1. Download to the destination all files this rank needs if on first replica
+            if first_replica:
+                log.debug(f'Rank {dist.get_global_rank()} starting to download files.')
+                for plan_item in plan.items:
+                    # Each plan item has a storage index which points to the relative path of the shard file at save time.
+                    relative_file_path = self.storage_data[plan_item.storage_index].relative_path
+                    # Download the shard file to the relative path it's associated to and save that relative path
+                    # to the root directory specified to the FileSystem reader constructor.
+                    file_destination = str(Path(self.destination_path) / Path(relative_file_path))
+                    # The file could have already been downloaded as different plan items can point to same file.
+                    if not os.path.exists(file_destination):
+                        log.debug(f'Downloading {relative_file_path} to {file_destination}.')
+                        self.object_store.download_object(object_name=str(
+                            Path(self.source_path) / Path(relative_file_path)),
+                                                          filename=file_destination)
+                        log.debug(f'Finished downloading {relative_file_path} to {file_destination}.')
 
             # 2. Wait for all ranks to finish.
+            log.debug(f'Finished downloading all files.')
             dist.barrier()
+            log.debug('Done waiting for all ranks to finish downloading files.')
+
+            # 3. Broadcast files to all other replicas if HSDP
+            if device_mesh is not None and device_mesh.ndim == 2:
+                # Broadcast file to all replicas. Assume replica size is at least 1 node
+                replicate_process_group = device_mesh.get_group(0)  # Replicate replicate_process_group
+                shard_size = device_mesh.size(1)
+
+                # Send list of files to all ranks
+                file_list = [sorted(os.listdir(self.destination_path))]
+                dist.broadcast_object_list(file_list,
+                                           src=dist.get_global_rank() % shard_size,
+                                           group=replicate_process_group)
+                file_list = file_list[0]
+                log.debug(f'{file_list=}')
+
+                # Send each file to the appropriate rank
+                for file_name in file_list:
+                    if 'metadata' in file_name:  # All ranks already have the metadata file
+                        continue
+                    if dist.get_local_rank() == 0:
```

Comment:
why is this local rank 0 gated? I would think that every rank in replica 0 needs to send its files to its peers?

Reply:
It looks in the filesystem which is shared across node, so only local rank 0 needs to do it.

Reply:
ahh i see. I think the variable names i requested below will make this more clear :)